src/apply.ts

```diff
@@ -26,154 +24,93 @@ interface ReadOptions {
 	application: number;
 	authToken: string;
 	stream: stream.Readable;
-	force?: boolean;
 }
 
-function generateUniqueKey() {
-	return uuidv4().replace(/-/g, '');
+interface Release {
+	releaseImages: Array<{
+		image: SDK.Image;
+		service: string;
+	}>;
 }
 
-async function backfillLowerRevisions(
-	sdk: SDK.BalenaSDK,
-	application: number,
-	currentDateIso: string,
-	manifest: ReleaseManifest,
-): Promise<void> {
-	const [highestSameSemverLocalReleases] = await sdk.pine.get({
-		resource: 'release',
-		options: {
-			$top: 1,
-			$select: ['id', 'revision'],
-			$filter: {
-				belongs_to__application: application,
-				semver_major: manifest.semver_major,
-				semver_minor: manifest.semver_minor,
-				semver_patch: manifest.semver_patch,
-				revision: { $ne: null },
-			},
-			$orderby: 'revision desc',
-		},
-	});
+function normalizeManifest(manifest: SDK.Release): Release {
+	const release: Release = { releaseImages: [] };
 
-	const version = `${manifest.semver_major}.${manifest.semver_minor}.${manifest.semver_patch}`;
-	for (
-		//
-		let revision =
-			highestSameSemverLocalReleases == null
-				? 0
-				: // we have already $filtered for this so ! is fine
-					highestSameSemverLocalReleases.revision! + 1;
-		revision < (manifest.revision ? manifest.revision : 0);
-		revision++
-	) {
-		await sdk.pine.post({
-			resource: 'release',
-			body: {
-				belongs_to__application: application,
-				status: 'failed',
-				start_timestamp: currentDateIso,
-				end_timestamp: currentDateIso,
-				update_timestamp: currentDateIso,
-				source: manifest.source,
-				composition: manifest.composition,
-				commit: generateUniqueKey(),
-				contract: manifest.contract,
-				semver: revision === 0 ? version : `${version}+rev${revision}`,
-			},
+	if (!Array.isArray(manifest.release_image)) {
+		throw new Error(
+			`Expected array of release images but found ${typeof manifest.release_image}`,
+		);
+	}
+	for (const releaseImage of manifest.release_image) {
+		const image = releaseImage.image;
+		if (!Array.isArray(image)) {
+			throw new Error(
+				`Expected array of images in release image ${releaseImage.id} but found ${typeof image}`,
```

Comment:
@dfunckt , how do you feel about this format of error message?

Reply:
It’s okay I suppose but overall we need to stop going in circles with irrelevant changes. The change makes sense in some cases but not in others (see my review comment below). If it wasn’t for this change, I could have approved and merged the pr and we could move on. We”re just spinning our wheels here.